pkg/canary/writer/push.go

```diff
@@ -185,24 +191,38 @@ func (p *Push) send(ctx context.Context, payload []byte) error {
 		req.SetBasicAuth(p.username, p.password)
 	}
 
-	resp, err := p.httpClient.Do(req)
-	if err != nil {
-		return fmt.Errorf("failed to push payload: %w", err)
-	}
-	defer func() {
-		if err := resp.Body.Close(); err != nil {
+	backoff := backoff.New(ctx, *p.backoff)
+	var resp *http.Response
+
+	// send log with retry
+	for {
+		resp, err = p.httpClient.Do(req)
+		if err != nil {
+			return fmt.Errorf("failed to push payload: %w", err)
+		}
+		status := resp.StatusCode
+
+		if status/100 != 2 {
+			scanner := bufio.NewScanner(io.LimitReader(resp.Body, defaultMaxReponseBufferLen))
+			line := ""
+			if scanner.Scan() {
+				line = scanner.Text()
+			}
+			err = fmt.Errorf("server returned HTTP status %s (%d): %s", resp.Status, status, line)
```

Comment:
Is this err logged anywhere?

Reply:
It is returned from this func at the end. And will be logged by the client calling `.Write()`.
https://github.com/grafana/loki/blob/kavirajk%2Fsupport-push-mode-canary/pkg/canary/writer/writer.go#L100

Reply:
Ah, gotcha. Thanks.